qiskit/providers/backend.py

```diff
@@ -229,3 +239,392 @@ class can handle either situation.
             Job: The job object for the run
         """
         pass
+
+
+class QubitProperties:
+    """A representation of the properties of a qubit on a backend.
+
+    This class provides the optional properties that a backend can provide for
+    a qubit. These represent the set of qubit properties that Qiskit can
+    currently work with if present. Howeverm if your backend provides additional
+    properties of qubits you should subclass this to add additional custom
+    attributes for those custom/additional properties provided by the backend.
+    """
+
+    __slots = ("t1", "t2", "frequency")
+
+    def __init__(self, t1=None, t2=None, frequency=None, properties=None):
+        """Create a new ``QubitProperties`` object
+
+        Args:
+            t1: The T1 time for a qubit in seconds
+            t2: The T2 time for a qubit in seconds
+            frequency: The frequency of a qubit in Hz
+            properties: A free form dictionary of additional properties the
+                backend has for a qubit.
+        """
+        self.t1 = t1
+        self.t2 = t2
+        self.frequency = frequency
+        self.properties = properties
+
+    def __repr__(self):
+        return (
+            f"QubitProperties(t1={self.t1}, t2={self.t2}, "
+            f"frequency={self.frequency}, properties={self.properties})"
+        )
+
+
+class BackendV2(Backend, ABC):
+    """Abstract class for Backends
+
+    This abstract class is to be used for all Backend objects created by a
+    provider. This version differs from earlier abstract Backend classes in
+    that the configuration attribute no longer exists. Instead, attributes
+    exposing equivalent required immutable properties of the backend device
+    are added. For example ``backend.configuration().n_qubits`` is accessible
+    from ``backend.num_qubits`` now.
+
+    The ``options`` attribute of the backend is used to contain the dynamic
+    user configurable options of the backend. It should be used more for
+    runtime options that configure how the backend is used. For example,
+    something like a ``shots`` field for a backend that runs experiments which
+    would contain an int for how many shots to execute.
+
+    If migrating a provider from :class:`~qiskit.providers.BackendV1` or
+    :class:`~qiskit.providers.BaseBackend` one thing to keep in mind is for
+    backwards compatibility you might need to add a configuration method that
+    will build a :class:`~qiskit.providers.models.BackendConfiguration` object
+    and :class:`~qiskit.providers.models.BackendProperties` from the attributes
+    defined in this class for backwards compatibility.
+    """
+
+    version = 2
```

Comment:
Is there any chance we're going to want minor versioning as well?  Again, I'm guessing no, but if so, it'll be easier to do the change now than later (e.g. to a tuple).

Reply:
No,  there shouldn't ever be a need for a minor version. When we introduced BackendV1 about a year ago we went with a monotonically increasing integer here because it's easy to reason about and we're doing supported ranges (like terra 0.20.0 will support backend.version >=1 <=2, while 0.19.0 is >=0, which is `BaseBackend`, <=2) there really isn't anything to indicate with a minor version. It's not semver and more like compatibility levels. We did document this in: https://qiskit.org/documentation/apidoc/providers.html#version-support (and the sections below).

Reply:
Perfect, I've got no problems with that at all.